source/agora/consensus/EnrollmentManager.d

```diff
@@ -577,9 +577,8 @@ public class EnrollmentManager
         foreach (const ref key; keys)
         {
             const preimage = this.validator_set.getPreimageAt(key, height);
-            // this should not happen. validators which didn't reveal the
```

Comment:
Please revert these changes, this assert is a preventative measure so the function is not called with the wrong height.

Reply:
I'm trying to restore the removed assertion, `assert(preimage != PreImageInfo.init);` by adding the code into the test codes like `QuorumPreimage.d` as follows.

```
foreach (enroll; enrolls)
{
    nodes.each!(node => retryFor(node.getPreimage(enroll.utxo_key).distance >= 1 , 10.seconds));
}
```

But, this code fails after outsider validators are enrolled. I guess it's because the outsider validators don't reveal their pre-images timely although they are enrolled as a validator. I need more debugging for that. How can I distinguish original validators from outsider validators in the source code? Is there any variable to distinguish them? Thanks in advance.

Reply:
```
foreach (enroll; blocks[0].header.enrollments)
{
    nodes.each!(node =>
    retryFor(node.getPreimage(enroll.utxo_key).distance >= 1 , 10.seconds));
}
```

With the codes like above, we can avoid assertions failure in many tests in `agora.test`. But, In QuorumPreimage.d, some failures occur continuously, which is related to outsiders I think. I’m debugging for that.

Reply:
Yes, this needs to be reverted. I can take a look.